crates/ra_syntax/src/validation.rs

```diff
@@ -1,40 +1,89 @@
+    string_lexing::{self, CharComponentKind},
     yellow::{
         SyntaxError,
     },
 };
 
 pub(crate) fn validate(file: &File) -> Vec<SyntaxError> {
     let mut errors = Vec::new();
-    for d in file.root.borrowed().descendants() {
-        if let Some(c) = ast::Char::cast(d) {
-            let components = &mut string_lexing::parse_char_literal(c.text());
-            let len = components.count();
+    let _ = visitor_ctx(&mut errors)
+        .visit::<ast::Char, _>(validate_char)
+        .accept(file.root.borrowed());
+    errors
+}
 
-            if !components.has_closing_quote {
-                errors.push(SyntaxError {
-                    msg: "Unclosed char literal".to_string(),
-                    offset: d.range().start(),
-                });
-            }
+fn validate_char(node: ast::Char, errors: &mut Vec<SyntaxError>) {
+    let mut components = string_lexing::parse_char_literal(node.text());
+    let mut len = 0;
+    for component in &mut components {
+        len += 1;
 
-            if len == 0 {
-                errors.push(SyntaxError {
-                    msg: "Empty char literal".to_string(),
-                    offset: d.range().start(),
-                });
+        // Validate escapes
+        // let text = &node.text()[component.range];
+        use self::CharComponentKind::*;
+        match component.kind {
+            AsciiEscape => {
+                // if text.len() == 1 {
+                //     // Escape sequence consists only of leading `\`
+                //     errors.push(SyntaxError {
+                //         msg: "Empty escape sequence".to_string(),
+                //         offset: component.range.start(),
+                //     });
+                // } else {
+                //     let escape_code = text.chars().skip(1).next().unwrap();
+                //     if !is_ascii_escape(escape_code) {
+                //         errors.push(SyntaxError {
+                //             msg: "Invalid escape sequence".to_string(),
+                //             offset: component.range.start(),
+                //         });
+                //     }
+                // }
             }
-
-            if len > 1 {
-                errors.push(SyntaxError {
-                    msg: "Character literal should be only one character long".to_string(),
-                    offset: d.range().start(),
-                });
+            AsciiCodeEscape => {
+                // TODO:
+                // * First digit is octal
+                // * Second digit is hex
             }
+            UnicodeEscape => {
+                // TODO:
+                // * Only hex digits or underscores allowed
+                // * Max 6 chars
+                // * Within allowed range (must be at most 10FFFF)
+            }
+            // Code points are always valid
+            CodePoint => ()
         }
     }
-    errors
+
+    if !components.has_closing_quote {
+        errors.push(SyntaxError {
```

Comment:
I think it makes sense to create a helper funciton for construction SyntaxError.